src/emqx_session.erl

```diff
@@ -634,16 +625,39 @@ replay(ClientInfo, Session = #session{inflight = Inflight}) ->
     end.
 
 -spec(terminate(emqx_types:clientinfo(), Reason :: term(), session()) -> ok).
-terminate(ClientInfo, discarded, Session) ->
+terminate(ClientInfo, Reason, Session) ->
+    run_terminate_hooks(ClientInfo, Reason, Session),
+    cleanup_self_from_shared_subs(),
+    redispatch_shared_messages(Session),
+    ok.
+
+run_terminate_hooks(ClientInfo, discarded, Session) ->
     run_hook('session.discarded', [ClientInfo, info(Session)]);
-terminate(ClientInfo, takeovered, Session) ->
+run_terminate_hooks(ClientInfo, takeovered, Session) ->
     run_hook('session.takeovered', [ClientInfo, info(Session)]);
-terminate(ClientInfo, Reason, Session) ->
+run_terminate_hooks(ClientInfo, Reason, Session) ->
     run_hook('session.terminated', [ClientInfo, Reason, info(Session)]).
 
+redispatch_shared_messages(#session{inflight = Inflight}) ->
+    InflightList = emqx_inflight:to_list(Inflight),
+    lists:map(fun({_, {#message{topic = Topic} = Msg, _}}) ->
```

Comment:
There may be a function_clause crash if the `InflightList` contains a  `{pubrel, Now}` element.
That value updated  by `emqx_inflight:update(PacketId, with_ts(pubrel), Inflight)` of `pubrec/2` func